### PR TITLE
conda enviroment configuration for build and run qgis

### PR DESCRIPTION
**Pull request by jadsilva**
## Description

With a conda isolated enviroment, developers can build and run qgis without need of installing a lot of dependencies to SO, neither dealing with containers. They just need to install [miniforge](https://github.com/conda-forge/miniforge#install) locally and create the enviroment from file conda-deps.yml that I put here. Developers can add new dependencies to the enviroment updating the file e reloading it.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@SrNetoChan does this look ok to you?

**Comment by SrNetoChan** [comment]
@jadsilva can you explain the rationale on why we want this in here? 
CC @gillins

**Comment by jadsilva** [comment]
> @jadsilva can you explain the rationale on why we want this in here? CC @gillins

@SrNetoChan , I put it in description

**Comment by SrNetoChan** [comment]
@jadsilva for installing QGIS using conda, this yaml file is redundant. having conda, miniconda, etc... all you need is one command:

conda create -n qgis -c conda-forge qgis
This will create an environment with all the necessary dependencies and install qgis in it.

If the purpose is to help developers to build qgis using conda, then they need a bit more than just an environment, they need a actual recipe, like the one used by conda-forge https://github.com/conda-forge/qgis-feedstock/tree/main/recipe. So, I still don't see the point of having this environment yaml in QGIS source code.

**Comment by jadsilva** [comment]
> @jadsilva for installing QGIS using conda, this yaml file is redundant. having conda, miniconda, etc... all you need is one command:
> 
> conda create -n qgis -c conda-forge qgis This will create an environment with all the necessary dependencies and install qgis (latest versio) in it.
> 
> If the purpose is to help developers to build qgis using conda, then they need a bit more than just an environment, they need a actual recipe, like the one used by conda-forge https://github.com/conda-forge/qgis-feedstock/tree/main/recipe. So, I still don't see the point of having this environment yaml in QGIS source code.

@SrNetoChan , this environment yaml has dependencies plus build tools. Devs can use it to create and activate a temporary and local env to build from source and run qgis without needing a recipe. It avoids installing all dependencies by using apt-get for example. I have tried to build and run qgis from source but I dont wanna polute my SO with all deps, and building/running using container is very uncomfortable, so I got this using this yml for creating a env. I have been using this strategy in other c++ projects and it has been working well.

**Comment by SrNetoChan** [comment]
In that case, more context is needed. For example how to build qgis using that environment. The instructions for installation of qgis that you have in the top of the yaml file will not install the current master, but the latest stable version available on conda forge.

Maybe expand the INSTALL.md file with instructions for that?

**Comment by jadsilva** [comment]
> In that case, more context is needed. For example how to build qgis using that environment. The instructions for installation of qgis that you have in the top of the yaml file will not install the current master, but the latest stable version available on conda forge.
> 
> Maybe expand the INSTALL.md file with instructions for that?

Yes, I'll

**Comment by gillins** [comment]
Would it be possible to add a CI/CD run for `conda`? GDAL has this, has been useful in detecting problems with different versions of libs etc before release...

**Review by DelazJ — COMMENTED** [review]
Hi. Just some typos and formatting suggestions...

File: INSTALL.md [hunk 1]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
 dnf install ccache ninja-build astyle python3-autopep8 python3-mock python3-nose2 perl-YAML-Tiny
 ```
 
+# 3.12. Building on local conda enviroment
```
[the remark this answers is not in the document]
Comment (jadsilva) [reply INSTALL.md]: Done

File: INSTALL.md [hunk 2]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
...
+
+With a conda isolated enviroment, developers can build and run qgis without need of installing a lot of dependencies to SO, neither dealing with containers. They just need to install [miniforge](https://github.com/conda-forge/miniforge#install) locally and create the enviroment from file conda-deps.yml. Developers can also add new dependencies to the enviroment updating the file e reloading it.
```
[the remark this answers is not in the document]
Comment (jadsilva) [reply INSTALL.md]: Done

File: INSTALL.md [hunk 3]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
...
+
+[Install miniforge](https://github.com/conda-forge/miniforge#install)
+
+[3.6. Check out the QGIS Source Code](#36-check-out-the-qgis-source-code)
```
[the remark this answers is not in the document]
Comment (jadsilva) [reply INSTALL.md]: done

File: INSTALL.md [hunk 4]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
...
 
+# 3.12. Building on local conda environment
+
+With a conda isolated environment, developers can build and run qgis without need of installing a lot of dependencies to O.S., neither dealing with containers. They just need to install [miniforge](https://github.com/conda-forge/miniforge#install) locally and create the environment from file conda-deps.yml. Developers can also add new dependencies to the environment updating the file and reloading it.
+
+1. [Install miniforge](https://github.com/conda-forge/miniforge#install)
+
+2. [Check out the QGIS Source Code](#36-check-out-the-qgis-source-code)
+
+3. Create or update the environment:
```
Comment (SrNetoChan) [inline INSTALL.md:559]:
```suggestion
3. Create the environment:
```

Comment (SrNetoChan) [reply INSTALL.md:559]: Let's focus on the install first.

File: INSTALL.md [hunk 5]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
...
+
+```bash
+mamba env update -n qgis -f conda-deps.yml
```
Comment (SrNetoChan) [inline INSTALL.md:562]:
```suggestion
mamba create -n qgis -f conda-deps.yml
```
AFAIK the env command have been deprecated

File: INSTALL.md [hunk 6]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
...
+```
+
+4. Use the environment:
+
+```bash
+conda activate qgis
+```
+
+5. [Starting the compile](#37-starting-the-compile)
+
```
Comment (SrNetoChan) [inline INSTALL.md:572]:
```suggestion
6. For updating the environment using a new conda-deps.yml:

        mamba update -n qgis -f conda-deps.yml
```

**Review by SrNetoChan — COMMENTED** [review]
I have some suggestions. One question, which I think is important. Will you keep this environment file updated in the future?

**Comment by jadsilva** [comment]
> I have some suggestions. One question, which I think is important. Will you keep this environment file updated in the future?

Any develop will can. In the same way they update dependencies in documentation

File: INSTALL.md [hunk 7]
```diff
@@ -547,6 +548,28 @@ are required for the various QGIS source formatting and preparation scripts.
 dnf install ccache ninja-build astyle python3-autopep8 python3-mock python3-nose2 perl-YAML-Tiny
 ```
 
+# 3.12. Building on local conda environment
+
+With a conda isolated environment, developers can build and run qgis without need of installing a lot of dependencies to O.S., neither dealing with containers. They just need to install [miniforge](https://github.com/conda-forge/miniforge#install) locally and create the environment from file conda-deps.yml. Developers can also add new dependencies to the environment updating the file and reloading it.
+
+1. [Install miniforge](https://github.com/conda-forge/miniforge#install)
```
[the remark this answers is not in the document]
Comment (jadsilva) [reply INSTALL.md:555]: https://github.com/conda-forge/miniforge#whats-the-difference-between-mambaforge-and-miniforge

**Comment by SrNetoChan** [comment]
> > I have some suggestions. One question, which I think is important. Will you keep this environment file updated in the future?
> 
> Any developer will can. In the same way they update dependencies in documentation

My question was not that. When we contribute or add code to a repository, we need to care about its maintenance. 
If no one updates the environment file, it will become useless and just cluttering the repository with non relevant information. So my question was if you intended to keep updating this env file? At least in the near future.

**Comment by jadsilva** [comment]
> > > I have some suggestions. One question, which I think is important. Will you keep this environment file updated in the future?
> > 
> > 
> > Any developer will can. In the same way they update dependencies in documentation
> 
> My question was not that. When we contribute or add code to a repository, we need to care about its maintenance. If no one updates the environment file, it will become useless and just cluttering the repository with non relevant information. So my question was if you intended to keep updating this env file? At least in the near future.

Yes, I Will. Its important too to have a CI/CD run for conda env, how @gillins https://github.com/qgis/QGIS/pull/55335#issuecomment-1825035112

File: INSTALL.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (SrNetoChan) [reply INSTALL.md:555]: ah got it.

**Review by SrNetoChan — APPROVED** [review]

**Comment by SrNetoChan** [comment]
> > > > I have some suggestions. One question, which I think is important. Will you keep this environment file updated in the future?
> > > 
> > > 
> > > Any developer will can. In the same way they update dependencies in documentation
> > 
> > 
> > My question was not that. When we contribute or add code to a repository, we need to care about its maintenance. If no one updates the environment file, it will become useless and just cluttering the repository with non relevant information. So my question was if you intended to keep updating this env file? At least in the near future.
> 
> Yes, I Will. Its important too to have a CI/CD run for conda env, how @gillins [#55335 (comment)](https://github.com/qgis/QGIS/pull/55335#issuecomment-1825035112)

I am pretty sure it would be possible, but not sure how the project sees it as there are already so many CI running. 
I value the opinion of @nyalldawson  @3nids and @m-kuhn on that subject

We would also need someone to implement it.

**Comment by m-kuhn** [comment]
Sorry, for the slow response.

I welcome instructions for building with other buildsystems like conda.

I am however not completely convinced that we should add code or configuration in this repository.

For CI code, it's a question of what does it bring more: noise because of occasional failures or benefits of discovering issues in QGIS code? Given that we already have a very broad CI coverage over plenty of scenarios, I think we already discover a broad set of _our_ issues, so I am not sure that adding more to the mix will add for QGIS code quality. I would say that if we add something new, we better let go of something else.

As for the environment configuration, I don't mind too much adding this, especially as we have someone volunteering to keeping this up to date (thanks @jadsilva). However, I wonder, if it wouldn't be possible to just document the steps for building with the an environment taken from the conda-forge QGIS repository, all the dependency are already maintained there, isn't this just duplicating this work?

**Comment by gillins** [comment]
I would just say that we carry a bunch of conda forge [patches](https://github.com/conda-forge/qgis-feedstock/tree/main/recipe) to workaround various things in QGIS. [At least one of these](https://github.com/conda-forge/qgis-feedstock/blob/main/recipe/0005-check-locks-not-destroyed.patch) is a clear bug in QGIS which somehow never gets triggered with the other combinations of compilers/dependency versions that the existing CI runs have.

Having these caught by a QGIS CI when the change is made would save us a lot of trouble. And potentially other QGIS packagers too when they upgrade to the same versions we use. No idea what to let go of instead though....

**Comment by m-kuhn** [comment]
It would be great if you could upstream those patches, so you don't have to maintain them any longer 🙏

For some context regarding our current troubles see https://github.com/qgis/QGIS/pull/55547#issuecomment-1877106773

**Comment by gregorywaynepower** [comment]
@jadsilva I wanted to chime in and agree that this _could_ be helpful for managing and building dependencies. However I'd recommend that you'd use [conda-lock](https://github.com/conda/conda-lock) when constructing the file that lists your dependencies since it hashes and points to the particular repositories you're pulling from, along with trying to resolve dependencies for multiple build targets in one fell swoop.

**Pull request closed without merging**